test/model/typing_status_test.dart

```diff
@@ -0,0 +1,240 @@
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
+    });
+
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.otherUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(dmNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.otherUser.userId: anyTimer}});
+      checkNotNotified();
+    }));
+
+    test('remove one of two typists in the same narrow', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.otherUser);
+      checkNotifiedOnce();
+      checkTyping(groupNarrow, TypingOp.start, eg.selfUser);
+      checkNotifiedOnce();
+
+      check(model).typistsByNarrow.deepEquals({
+        groupNarrow: {eg.otherUser.userId: anyTimer, eg.selfUser.userId: anyTimer}});
+
+      checkTyping(groupNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    }));
+
+    test('remove typists from different narrows', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(dmNarrow, TypingOp.start, eg.otherUser);
+      checkNotifiedOnce();
+      checkTyping(groupNarrow, TypingOp.start, eg.thirdUser);
+      checkNotifiedOnce();
+      checkTyping(topicNarrow, TypingOp.start, eg.selfUser);
+      checkNotifiedOnce();
+
+      async.elapse(const Duration(seconds: 5));
+      check(model).typistsByNarrow.deepEquals({
+        dmNarrow: {eg.otherUser.userId: anyTimer},
+        groupNarrow: {eg.thirdUser.userId: anyTimer},
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+
+      checkTyping(groupNarrow, TypingOp.stop, eg.thirdUser)
+        .typistsByNarrow.deepEquals({
+          dmNarrow: {eg.otherUser.userId: anyTimer},
+          topicNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(dmNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.deepEquals({topicNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(topicNarrow, TypingOp.stop, eg.selfUser)
+        .typistsByNarrow.isEmpty();
+      checkNotifiedOnce();
+    }));
+  });
+
+  group('cancelling old timer', () {
+    test('when typing stopped early', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(dmNarrow, TypingOp.start, eg.otherUser);
+      final oldTimer = check(model).typistsByNarrow[dmNarrow][eg.otherUser.userId]
+        ..isActive.isTrue();
+      checkNotifiedOnce();
+
+      checkTyping(dmNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.isEmpty();
+      oldTimer.isActive.isFalse();
+      checkNotifiedOnce();
+    }));
```

Comment:
Here's a bit more comprehensive of a check:
```suggestion
      checkNotifiedOnce();
      check(async.pendingTimers).isEmpty();
    }));
```

In fact I probably should adjust `awaitFakeAsync` so that it automatically checks that `async.pendingTimers` is empty when the callback (i.e. the test body) completes. If a test accidentally leaks timers, that's a fairly likely way of learning about a bug in the code under test.

That also avoids the need to keep a `Subject<Timer>` around, in that variable `oldTimer`. Doing that (as opposed to immediately invoking a series of methods on the `Subject` and then discarding it) is something I don't think is conventional, and I'm not entirely sure how it behaves — `Subject` is a bit opaque that way.